pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1043,7 +1047,15 @@ impl Timeline {
             // We have a list of deltas/images. We want to create image layers while collect garbages.
             for (key, lsn, val) in accumulated_values.iter().rev() {
                 if *lsn > horizon {
-                    keys_above_horizon.push((*key, *lsn, val.clone())); // TODO: ensure one LSN corresponds to either delta or image instead of both
+                    if let Some((_, prev_lsn, _)) = keys_above_horizon.last_mut() {
+                        if *prev_lsn == *lsn {
+                            // The case that we have an LSN with both data from the delta layer and the image layer. As
+                            // `ValueWrapper` ensures that an image is ordered before a delta at the same LSN, we simply
+                            // drop this delta and keep the image.
+                            continue;
```

Comment:
Does this happen in real life or just in your test cases?

Reply:
This happens in real when we create an image layer at the end LSN of a delta layer.

Reply:
Maybe the comment should mention that.

Reply:
added in https://github.com/neondatabase/neon/pull/8078/commits/b139c895755188ae7abc4852c88434ff91f83b43